File: app/src/main/java/com/androidevlinux/percy/UTXO/adapters/GlobalPriceAdapter.java
package com.androidevlinux.percy.UTXO.adapters;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.transition.ChangeBounds;
import android.support.transition.TransitionManager;
import android.support.v7.widget.AppCompatImageView;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.androidevlinux.percy.UTXO.R;
import com.androidevlinux.percy.UTXO.data.models.coinmarketcap.CoinMarketCapCoin;
import com.squareup.picasso.Picasso;

import java.text.MessageFormat;
import java.util.ArrayList;

public class GlobalPriceAdapter extends RecyclerView.Adapter<GlobalPriceAdapter.ViewHolder> {

    private ArrayList<CoinMarketCapCoin> coinMarketCapCoinArrayList;
    private Context context;
    private RecyclerView recyclerView;
    private int mExpandedPosition = -1;

    public GlobalPriceAdapter(ArrayList<CoinMarketCapCoin> coinMarketCapCoinArrayList, Context context, RecyclerView recyclerView) {
        this.context = context;
        this.recyclerView = recyclerView;
        this.coinMarketCapCoinArrayList = coinMarketCapCoinArrayList;
    }

    @NonNull
    @Override
    public GlobalPriceAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.global_price_fragment_adapter_row, parent, false);
        return new GlobalPriceAdapter.ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull final GlobalPriceAdapter.ViewHolder holder, int position) {
        CoinMarketCapCoin coinMarketCapCoin = coinMarketCapCoinArrayList.get(position);
        holder.id.setText(String.valueOf(position + 1));
        holder.title.setText(coinMarketCapCoin.getData().getName());
        holder.price.setText(MessageFormat.format("{0}{1}", context.getResources().getString(R.string.price) + " $", coinMarketCapCoin.getData().getQuotes().getUSD().getPrice()));
        holder.txt_volume_24h.setText(MessageFormat.format("{0}{1}", context.getResources().getString(R.string.volume24h) + "$", coinMarketCapCoin.getData().getQuotes().getUSD().getVolume24h()));

        holder.txt_1hr_percent_change.setText(MessageFormat.format("{0}{1}", coinMarketCapCoin.getData().getQuotes().getUSD().getPercentChange1h(), "% 1Hr"));
        if (coinMarketCapCoin.getData().getQuotes().getUSD().getPercentChange1h() > 0) {
            holder.txt_1hr_percent_change.setTextColor(context.getResources().getColor(R.color.md_green_900));
        } else {
            holder.txt_1hr_percent_change.setTextColor(context.getResources().getColor(R.color.md_red_900));
        }

        holder.txt_24hr_percent_change.setText(MessageFormat.format("{0}{1}", coinMarketCapCoin.getData().getQuotes().getUSD().getPercentChange24h(), "% 24Hr"));
        if (coinMarketCapCoin.getData().getQuotes().getUSD().getPercentChange24h() > 0) {
            holder.txt_24hr_percent_change.setTextColor(context.getResources().getColor(R.color.md_green_900));
        } else {
            holder.txt_24hr_percent_change.setTextColor(context.getResources().getColor(R.color.md_red_900));
        }

        holder.txt_7d_percent_change.setText(MessageFormat.format("{0}{1}", coinMarketCapCoin.getData().getQuotes().getUSD().getPercentChange7d(), "% 7d"));
        if (coinMarketCapCoin.getData().getQuotes().getUSD().getPercentChange7d() > 0) {
            holder.txt_7d_percent_change.setTextColor(context.getResources().getColor(R.color.md_green_900));
        } else {
            holder.txt_7d_percent_change.setTextColor(context.getResources().getColor(R.color.md_red_900));
        }

        Picasso.get().load("https://s2.coinmarketcap.com/generated/sparklines/web/7d/usd/" + String.valueOf(coinMarketCapCoin.getData().getId()) + ".png").into(holder.snapshotImage);
        Picasso.get().load("https://s2.coinmarketcap.com/static/img/coins/64x64/" + String.valueOf(coinMarketCapCoin.getData().getId()) + ".png").into(holder.exchangeImage);
        final boolean isExpanded = position == mExpandedPosition;
        holder.expandedView.setVisibility(isExpanded ? View.VISIBLE : View.GONE);
        holder.itemView.setActivated(isExpanded);
        holder.itemView.setOnClickListener(v -> {
            boolean shouldExpand = holder.expandedView.getVisibility() == View.GONE;
            mExpandedPosition = isExpanded ? -1 : holder.getAdapterPosition();
            ChangeBounds transition = new ChangeBounds();
            transition.setDuration(125);

            if (shouldExpand) {
                holder.expandedView.setVisibility(View.VISIBLE);
                holder.imageView_toggle.setImageResource(R.drawable.ic_expand_less);
            } else {
                holder.expandedView.setVisibility(View.GONE);
                holder.imageView_toggle.setImageResource(R.drawable.ic_expand_more);
            }

            TransitionManager.beginDelayedTransition(recyclerView, transition);
            holder.itemView.setActivated(shouldExpand);
        });
    }


    @Override
    public int getItemCount() {
        return coinMarketCapCoinArrayList.size();
    }

    /**
     * View holder class
     */
    class ViewHolder extends RecyclerView.ViewHolder {
        private CardView cardView;
        private TextView id;
        private AppCompatImageView exchangeImage;
        private TextView title;
        private TextView price;
        private TextView txt_volume_24h;
        private TextView txt_1hr_percent_change;
        private TextView txt_24hr_percent_change;
        private TextView txt_7d_percent_change;
        private AppCompatImageView snapshotImage;
        private LinearLayout expandedView;
        private ImageView imageView_toggle;

        private ViewHolder(View view) {
            super(view);
            cardView = view.findViewById(R.id.cardView);
            expandedView = view.findViewById(R.id.expandView);
            imageView_toggle = view.findViewById(R.id.imageView_toggle);
            id = view.findViewById(R.id.id);
            exchangeImage = view.findViewById(R.id.exchangeImage);
            title = view.findViewById(R.id.txt_title);
            price = view.findViewById(R.id.txt_price);
            txt_volume_24h = view.findViewById(R.id.txt_volume_24h);
            txt_1hr_percent_change = view.findViewById(R.id.txt_1hr_percent_change);
            txt_24hr_percent_change = view.findViewById(R.id.txt_24hr_percent_change);
            txt_7d_percent_change = view.findViewById(R.id.txt_7d_percent_change);
            snapshotImage = view.findViewById(R.id.snapshotImage);
        }
    }
}

File: app/src/main/java/com/androidevlinux/percy/UTXO/adapters/GlobalPriceAdapter.kt
package com.androidevlinux.percy.UTXO.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.appcompat.widget.AppCompatImageView
import androidx.core.content.ContextCompat
import androidx.swiperefreshlayout.widget.CircularProgressDrawable
import com.androidevlinux.percy.UTXO.R
import com.androidevlinux.percy.UTXO.data.models.coinmarketcap.CoinMarketCapCoin
import com.squareup.picasso.Picasso
import java.text.MessageFormat
import java.util.*


class GlobalPriceAdapter(private val coinMarketCapCoinArrayList: ArrayList<CoinMarketCapCoin>, private val context: Context, private val mSwipeRefreshLayout: androidx.swiperefreshlayout.widget.SwipeRefreshLayout) : androidx.recyclerview.widget.RecyclerView.Adapter<GlobalPriceAdapter.ViewHolder>() {
    private var lastPosition = -1

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GlobalPriceAdapter.ViewHolder {
        val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.global_price_fragment_adapter_row, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: GlobalPriceAdapter.ViewHolder, position: Int) {
        val coinMarketCapCoin = coinMarketCapCoinArrayList[position]
        val animation = AnimationUtils.loadAnimation(
                context,
                if (position > lastPosition)
                    R.anim.up_from_bottom
                else
                    R.anim.down_from_top
        )
        holder.itemView.startAnimation(animation)
        lastPosition = holder.adapterPosition

        holder.id.text = (position + 1).toString()
        holder.title.text = coinMarketCapCoin.data!!.name
        holder.price.text = MessageFormat.format("{0}{1}", context.resources.getString(R.string.price) + " $", coinMarketCapCoin.data!!.quotes!!.usd!!.price)
        holder.txtVolume24h.text = MessageFormat.format("{0}{1}", context.resources.getString(R.string.volume24h) + "$", coinMarketCapCoin.data!!.quotes!!.usd!!.volume24h)

        holder.txt1hrPercentChange.text = MessageFormat.format("{0}{1}", coinMarketCapCoin.data!!.quotes!!.usd!!.percentChange1h, "% 1Hr")
        if (coinMarketCapCoin.data!!.quotes!!.usd!!.percentChange1h!! > 0) {
            holder.txt1hrPercentChange.setTextColor(ContextCompat.getColor(context, R.color.md_green))
        } else {
            holder.txt1hrPercentChange.setTextColor(ContextCompat.getColor(context, R.color.md_red))
        }

        holder.txt24hrPercentChange.text = MessageFormat.format("{0}{1}", coinMarketCapCoin.data!!.quotes!!.usd!!.percentChange24h!!, "% 24Hr")
        if (coinMarketCapCoin.data!!.quotes!!.usd!!.percentChange24h!! > 0) {
            holder.txt24hrPercentChange.setTextColor(ContextCompat.getColor(context, R.color.md_green))
        } else {
            holder.txt24hrPercentChange.setTextColor(ContextCompat.getColor(context, R.color.md_red))
        }

        holder.txt7dPercentChange.text = MessageFormat.format("{0}{1}", coinMarketCapCoin.data!!.quotes!!.usd!!.percentChange7d, "% 7d")
        if (coinMarketCapCoin.data!!.quotes!!.usd!!.percentChange7d!! > 0) {
            holder.txt7dPercentChange.setTextColor(ContextCompat.getColor(context, R.color.md_green))
        } else {
            holder.txt7dPercentChange.setTextColor(ContextCompat.getColor(context, R.color.md_red))
        }

        val circularProgressDrawable = CircularProgressDrawable(context)
        circularProgressDrawable.strokeWidth = 5f
        circularProgressDrawable.centerRadius = 30f
        circularProgressDrawable.setColorSchemeColors(R.color.md_blue,
                R.color.md_pink)
        circularProgressDrawable.start()
        Picasso.get().load("https://s2.coinmarketcap.com/generated/sparklines/web/7d/usd/" + coinMarketCapCoin.data!!.id.toString() + ".png").placeholder(circularProgressDrawable).into(holder.snapshotImage)
        Picasso.get().load("https://s2.coinmarketcap.com/static/img/coins/64x64/" + coinMarketCapCoin.data!!.id.toString() + ".png").placeholder(circularProgressDrawable).into(holder.exchangeImage)

        // EOS is having dark black border icon which is not visible
        // in black background so add a white background color
        if (coinMarketCapCoin.data!!.id == 1765 || coinMarketCapCoin.data!!.id == 52 || coinMarketCapCoin.data!!.id == 1720 || coinMarketCapCoin.data!!.id == 1027) {
            holder.exchangeImage.setBackgroundColor(ContextCompat.getColor(context, android.R.color.white))
        } else {
            holder.exchangeImage.setBackgroundColor(ContextCompat.getColor(context, android.R.color.transparent))
        }

        val isExpanded = coinMarketCapCoin.isExpanded
        holder.expandedView.visibility = if (isExpanded) View.VISIBLE else View.GONE
        if (isExpanded) {
            holder.imageViewToggle.setImageResource(R.drawable.ic_expand_less)
        } else {
            holder.imageViewToggle.setImageResource(R.drawable.ic_expand_more)
        }
        holder.itemView.setOnClickListener {
            val isViewExpanded = coinMarketCapCoin.isExpanded
            if (isViewExpanded) {
                holder.expandedView.visibility = View.GONE
                coinMarketCapCoin.isExpanded = false
                holder.imageViewToggle.setImageResource(R.drawable.ic_expand_more)
            } else {
                holder.expandedView.visibility = View.VISIBLE
                coinMarketCapCoin.isExpanded = true
                holder.imageViewToggle.setImageResource(R.drawable.ic_expand_less)
            }
        }
        if (coinMarketCapCoinArrayList.size == 13) {
            if (mSwipeRefreshLayout.isRefreshing) {
                mSwipeRefreshLayout.isRefreshing = false
            }
        }
    }


    override fun getItemCount(): Int {
        return coinMarketCapCoinArrayList.size
    }

    /**
     * View holder class
     */
    inner class ViewHolder(view: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(view) {
        val cardView: androidx.cardview.widget.CardView = view.findViewById(R.id.cardView)
        val id: TextView = view.findViewById(R.id.id)
        val exchangeImage: AppCompatImageView = view.findViewById(R.id.exchangeImage)
        val title: TextView = view.findViewById(R.id.txt_title)
        val price: TextView = view.findViewById(R.id.txt_price)
        val txtVolume24h: TextView = view.findViewById(R.id.txt_volume_24h)
        val txt1hrPercentChange: TextView = view.findViewById(R.id.txt_1hr_percent_change)
        val txt24hrPercentChange: TextView = view.findViewById(R.id.txt_24hr_percent_change)
        val txt7dPercentChange: TextView = view.findViewById(R.id.txt_7d_percent_change)
        val snapshotImage: AppCompatImageView = view.findViewById(R.id.snapshotImage)
        val expandedView: RelativeLayout = view.findViewById(R.id.expandView)
        val mainLayout: RelativeLayout = view.findViewById(R.id.mainLayout)
        val imageViewToggle: ImageView = view.findViewById(R.id.imageView_toggle)

    }
}

